Update activiation fusion in OpenCL backend

leejaymin opened the pull request:
Summary:

[PR4189](https://github.com/pytorch/glow/pull/4189) provides Conv and Relu operator fusion for better performance.
Unfortunately, I found a problem that Conv and Relu nodes are not fused in case of quantization.

Case 1: Quantization node prevents Conv and Relu operator fusion.
- As shown in the below graph, when quantization and de-quantization nodes are located in the middle of Conv and Relu nodes. `foldActiviations` is not working properly. It is because the parent Relu node is not Conv node. Therefore, After `transformForPrecisionMode` pass, `foldActiviations` is not working. 
I think that the right position of `foldActiviations` is after `optimize()`. `optimize()` pass removes all quantization nodes. So, fusion properly occurs.



![image](https://user-images.githubusercontent.com/1924507/91724971-fe4ace00-ebd8-11ea-8edb-370fe07e8680.png)

Case 2: Need to keep the right scale factors
- As shown in the below graph, If Relu has different scales between **in(0.0625)** and **out(0.0312)**, `replaceAllUsesOfWith` function is not working. `Assert` in `replaceAllUsesOfWith` function raises an error due to below codes.
```c++
if (v.getNode() && getType() != v.getType()) {
    // Fresh replacement nodes without users are usually created by the graph
    // optimizer, where the type of the replacement value should always match
    // the type of the original node value. Check if the only difference is
    // related to strides and adjust accordingly.
    assert(v.getNumUsers() == 0 && "Cannot update type if there are users");
    assert(getType()->isEqual(*v.getType(), /* allowDifferentShape */ false,
                              /* allowDifferentStrides */ true) &&
           "Replacing value with the wrong type");
    v.setType(getType());
  }
```
To avoid this `assert` error and get proper results, the **out-scale(0.0625) of Conv node** should be replaced by **out-scale(0.0312) of Relu node** rather than **0.0625** before calling `replaceAllUsesOfWith` function.

**Problem Case** 
![image](https://user-images.githubusercontent.com/1924507/91725586-ecb5f600-ebd9-11ea-8f0d-99483d786444.png)
**Right Fusion**
![image](https://user-images.githubusercontent.com/1924507/91939300-8b546b00-ed30-11ea-8f22-7508e43835cb.png)


Documentation:

[Optional Fixes #issue]

Test Plan:
Added two test cases in BackendCorrectnessTest.

Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.

leejaymin commented:
@pjaaskel @jfix71 Please take a look. Thanks.

leejaymin commented:
@jfix71 Could you take a look ?
thanks.

jfix71 commented on lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:
```diff
@@ -5119,18 +5134,18 @@ Error glow::optimizeFunction(Function *F, const Backend &B,
+  // Fold activations after optimization.
+  // Quantization nodes should be optimized before folding Activation in to
+  // Conv.
+  foldActivations(F, cctx, &B);
```
Instead of moving this down here, maybe it could be left where it was, and we could add another call to `optimize()`? I.e. we'd have `transformForPrecisionMode()`, then `foldActivations()`, then `lower()`. Otherwise here we will only fold activations after lowering, which as the comment that was originally there states, makes pattern matching for activation folding a little more annoying.

leejaymin replied:
Sure. I moved `foldActivation` in the original position. Also, I added another `optimize()` before calling `foldActivation`.

jfix71 commented on lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:
```diff
@@ -4816,7 +4816,22 @@ template <class T> bool fuseActivation(T *N, Function *F, const Backend *B) {
   }
 
+  // currently not to support asymmetric quantization fusion.
+  if (N->getResult().getType()->isQuantizedType()) {
+    if ((N->getResult().getType()->getOffset() != 0) ||
+        (activationNV.getType()->getOffset() != 0)) {
```
```suggestion
  if (N->getResult().getType()->isQuantizedType() &&
      ((N->getResult().getType()->getOffset() != 0) ||
        (activationNV.getType()->getOffset() != 0)) {
```

leejaymin replied:
Change it.

jfix71 reviewed (commented):
@leejaymin Sorry for the delay, added a few comments. Also can you add a unit test that would catch the bug for quantization?

leejaymin commented:
@jfix71 Thank you for the code review. As you requested, I changed the code and added two test cases in `BackendCorrectnessTest` to verify the fusion for quantization in OpenCL backend. 
Please take a look again.

---

By the way, I work for [ETRI](https://www.etri.re.kr/eng/main/main.etri), funded by the South Korea government. At ETRI, our team and I are doing to implement the AI compiler by extending the Glow for **our open-AI accelerator** (like VTA from `TVM` community). With the Glow extension, we have also implemented several further optimizations, including **1)auto-tuning mixed-precision quantization**, **2)One-shot Inference-based Channel Pruning**, **3)Profiling-based Dynamic Partitioning**, and **4)Profiling-based dynamic tile scheduling**.

Similar to the Glow compiler, all of our works will be opened in accordance with the opensource license by the end of this year. In light of our purpose, we think that it might be good to talk together in some ways. **if the Glow team might be interested in our project, we look forward to talking together for partnership on this topic in more detail. We wonder that there is a way to get in touch with the Glow team such as the director's email. Please let me know via my email address which is <leejaymin@etri.re.kr>.**

leejaymin commented:
@jfix71 I hope to get my code reviewed. Thanks.

jfix71 reviewed (commented):
Will review tomorrow. Sorry about all the delays.

jfix71 commented on tests/unittests/BackendCorrectnessTest.cpp:
```diff
@@ -535,4 +535,44 @@ TEST_P(BackendCorrectnessTest, maxSplatTest) {
+  kernel.getHandle<int8_t>().randomize(-1, 1, PRNG);
+  bias.getHandle<int32_t>().randomize(0, 32768, PRNG);
+  std::array<dim_t, 4> S{{1, 6, 6, 1}};
+  llvm::ArrayRef<dim_t> shape(S);
+  Tensor out(ElemKind::Int8QTy, shape, 256, 0);
```
nit: We should be able to just pass S into the Tensor constructor, no? Arrays should implicitly get converted to ArrayRefs.

leejaymin replied:
I changed it directly to convey the shape of the tensor to the constructor.

jfix71 commented on tests/unittests/BackendCorrectnessTest.cpp:
```diff
@@ -535,4 +535,44 @@ TEST_P(BackendCorrectnessTest, maxSplatTest) {
+  llvm::ArrayRef<dim_t> shape(S);
+  Tensor out(ElemKind::Int8QTy, shape, 256, 0);
+
+  int ResFusion =
```
nit: `resFusion`

leejaymin replied:
Changed it.

jfix71 commented on tests/unittests/BackendCorrectnessTest.cpp:
```diff
@@ -535,4 +535,44 @@ TEST_P(BackendCorrectnessTest, maxSplatTest) {
+/// Asymmetric Quantized Conv+Relu fusion testing for only OpenCL.
+TEST_P(BackendCorrectnessTest, AsymmetricQuantizedConvReluFusionTest) {
+  CHECK_IF_ENABLED()
+  PseudoRNG PRNG;
+  Tensor inputs(ElemKind::Int8QTy, {1, 6, 6, 16}, 1, 2);
+  Tensor kernel(ElemKind::Int8QTy, {1, 3, 3, 16}, 1, 2);
+  Tensor bias(ElemKind::Int32QTy, {1}, 1, 2);
+  inputs.getHandle<int8_t>().randomize(0, 60, PRNG);
+  kernel.getHandle<int8_t>().randomize(-1, 1, PRNG);
+  bias.getHandle<int32_t>().randomize(0, 32768, PRNG);
+  std::array<dim_t, 4> S{{1, 6, 6, 1}};
+  llvm::ArrayRef<dim_t> shape(S);
+  Tensor out(ElemKind::Int8QTy, shape, 256, 4);
+
+  int ResFusion =
+      inferConvReluNet(&inputs, &kernel, &bias, &out, 3, 1, 1, backendName_);
```
nit: All of this logic is nearly the same as in `SymmetricQuantizedConvReluFusionTest` -- should be easy to create a helper that contains all of this logic which takes a bool for whether to use symmetric or not and set offsets accordingly.

leejaymin replied:
Thank you for the comment. As you recommended, I created the helper function.

jfix71 reviewed (approved):
LGTM, a few small nits. Sorry for the delay. Thanks for adding the test!

leejaymin commented:
Thank you for the constructive feedback. If it is fine, could we have this landed?
